File: src/engine/deck.ts
// src/engine/deck.ts — a pure, seeded draw/discard deck primitive.
// No React, DOM, I/O, stores, or services. No wall clock, no Math.random, no crypto.
// All randomness is derived from a numeric `seed` argument via shuffle from ./rng.

import { shuffle } from './rng';

/**
 * A draw/discard deck. JSON-serializable: only arrays + optional plain value.
 * `current` is the most recently activated item (e.g. the card "in play"); omitted
 * when none is set.
 */
export interface DeckState<T = string> {
  drawPile: T[];
  discardPile: T[];
  current?: T;
}

/** Build a fresh deck: items shuffled into the draw pile (seeded), empty discard. */
export function create<T>(items: T[], seed: number): DeckState<T> {
  return {
    drawPile: shuffle(items, seed),
    discardPile: [],
  };
}

export interface DrawResult<T> {
  deck: DeckState<T>;
  drawn: T[];
  reshuffled: boolean;
}

/**
 * Draw up to `n` items from the top of the draw pile (the end of the array).
 * When the draw pile is empty but the discard pile still has cards, the discard
 * pile is reshuffled (seeded) into the draw pile and `reshuffled` is true.
 *
 * Total: non-positive `n` is a no-op (returns the input deck unchanged with an
 * empty `drawn`). `n` larger than the total available cards draws everything.
 * Never mutates the input.
 */
export function draw<T>(s: DeckState<T>, n: number, seed: number): DrawResult<T> {
  const count = Math.floor(n);
  if (count <= 0) {
    return { deck: s, drawn: [], reshuffled: false };
  }

  let drawPile = s.drawPile.slice();
  let discardPile = s.discardPile.slice();
  let reshuffled = false;
  const drawn: T[] = [];

  for (let i = 0; i < count; i++) {
    if (drawPile.length === 0) {
      if (discardPile.length === 0) break; // truly exhausted
      // Reshuffle the discard pile back into the draw pile.
      drawPile = shuffle(discardPile, seed);
      discardPile = [];
      reshuffled = true;
    }
    drawn.push(drawPile.pop() as T);
  }

  const deck: DeckState<T> = { drawPile, discardPile };
  if ('current' in s) deck.current = s.current;

  return { deck, drawn, reshuffled };
}

/** Push an item onto the top of the discard pile. Returns a new deck. */
export function discard<T>(s: DeckState<T>, item: T): DeckState<T> {
  const deck: DeckState<T> = {
    drawPile: s.drawPile,
    discardPile: [...s.discardPile, item],
  };
  if ('current' in s) deck.current = s.current;
  return deck;
}

/** Set the `current` item. Returns a new deck (piles untouched). */
export function setCurrent<T>(s: DeckState<T>, item: T): DeckState<T> {
  return {
    drawPile: s.drawPile,
    discardPile: s.discardPile,
    current: item,
  };
}

/** Number of cards left to draw. */
export function remaining<T>(s: DeckState<T>): number {
  return s.drawPile.length;
}

/** True when neither pile has any cards left. */
export function isExhausted<T>(s: DeckState<T>): boolean {
  return s.drawPile.length === 0 && s.discardPile.length === 0;
}

/**
 * Merge both piles and shuffle (seeded) into the draw pile; discard becomes empty.
 * Preserves `current`. Returns a new deck.
 */
export function reshuffle<T>(s: DeckState<T>, seed: number): DeckState<T> {
  const deck: DeckState<T> = {
    drawPile: shuffle([...s.drawPile, ...s.discardPile], seed),
    discardPile: [],
  };
  if ('current' in s) deck.current = s.current;
  return deck;
}

File: android-native/app/src/main/java/com/gamenight/party/engine/Deck.kt
package com.gamenight.party.engine

/**
 * Port of src/engine/deck.ts — a pure, seeded draw/discard deck primitive.
 *
 * No clock, no Math.random: all randomness is derived from a numeric [seed] via [Rng.shuffle]
 * (a fresh [Rng] per call, exactly like the web `shuffle(items, seed)`). Operations never mutate
 * their input — they return new immutable state.
 */

/**
 * A draw/discard deck. [current] is the most recently activated item (the card "in play");
 * `null` means "none set" (the web's absent `current` key).
 */
data class DeckState<T>(
    val drawPile: List<T>,
    val discardPile: List<T>,
    val current: T? = null,
)

/** Result of a [draw]: the new [deck], the items [drawn], and whether a reshuffle happened. */
data class DrawResult<T>(
    val deck: DeckState<T>,
    val drawn: List<T>,
    val reshuffled: Boolean,
)

/** Build a fresh deck: items shuffled into the draw pile (seeded), empty discard. */
fun <T> create(items: List<T>, seed: Int): DeckState<T> =
    DeckState(drawPile = Rng(seed).shuffle(items), discardPile = emptyList())

/**
 * Draw up to [n] items from the top of the draw pile (the end of the list). When the draw pile is
 * empty but the discard pile still has cards, the discard pile is reshuffled (seeded) into the draw
 * pile and [DrawResult.reshuffled] is true.
 *
 * Total: non-positive [n] is a no-op (returns the input deck unchanged with an empty `drawn`).
 * [n] larger than the total available cards draws everything. Never mutates the input.
 */
fun <T> draw(s: DeckState<T>, n: Int, seed: Int): DrawResult<T> {
    if (n <= 0) return DrawResult(s, emptyList(), false)

    var drawPile = s.drawPile.toMutableList()
    var discardPile = s.discardPile.toMutableList()
    var reshuffled = false
    val drawn = ArrayList<T>()

    for (i in 0 until n) {
        if (drawPile.isEmpty()) {
            if (discardPile.isEmpty()) break // truly exhausted
            // Reshuffle the discard pile back into the draw pile.
            drawPile = Rng(seed).shuffle(discardPile).toMutableList()
            discardPile = ArrayList()
            reshuffled = true
        }
        drawn.add(drawPile.removeAt(drawPile.size - 1))
    }

    val deck = DeckState(drawPile.toList(), discardPile.toList(), s.current)
    return DrawResult(deck, drawn, reshuffled)
}

/** Push an item onto the top of the discard pile. Returns a new deck. */
fun <T> discard(s: DeckState<T>, item: T): DeckState<T> =
    DeckState(s.drawPile, s.discardPile + item, s.current)

/** Set the [DeckState.current] item. Returns a new deck (piles untouched). */
fun <T> setCurrent(s: DeckState<T>, item: T): DeckState<T> =
    DeckState(s.drawPile, s.discardPile, item)

/** Number of cards left to draw. */
fun <T> remaining(s: DeckState<T>): Int = s.drawPile.size

/** True when neither pile has any cards left. */
fun <T> isExhausted(s: DeckState<T>): Boolean =
    s.drawPile.isEmpty() && s.discardPile.isEmpty()

/**
 * Merge both piles and shuffle (seeded) into the draw pile; discard becomes empty.
 * Preserves [DeckState.current]. Returns a new deck.
 */
fun <T> reshuffle(s: DeckState<T>, seed: Int): DeckState<T> =
    DeckState(
        drawPile = Rng(seed).shuffle(s.drawPile + s.discardPile),
        discardPile = emptyList(),
        current = s.current,
    )
